File: swift-bridge/Sources/CoreFoundationBridge/Collections.swift
import CoreFoundation
import Foundation

private func acfArrayFromRawPointers(_ pointers: [UnsafeRawPointer?]) -> CFArray? {
    var callbacks = kCFTypeArrayCallBacks
    var mutablePointers = pointers
    return mutablePointers.withUnsafeMutableBufferPointer { buffer in
        CFArrayCreate(nil, buffer.baseAddress, buffer.count, &callbacks)
    }
}

private func acfDictionaryFromRawPointers(
    keys: [UnsafeRawPointer?],
    values: [UnsafeRawPointer?]
) -> CFDictionary? {
    precondition(keys.count == values.count)
    let count = keys.count
    var keys = keys
    var values = values
    var keyCallbacks = kCFTypeDictionaryKeyCallBacks
    var valueCallbacks = kCFTypeDictionaryValueCallBacks
    return keys.withUnsafeMutableBufferPointer { keyBuffer in
        values.withUnsafeMutableBufferPointer { valueBuffer in
            CFDictionaryCreate(
                nil,
                keyBuffer.baseAddress,
                valueBuffer.baseAddress,
                count,
                &keyCallbacks,
                &valueCallbacks
            )
        }
    }
}

private func acfBagFromRawPointers(_ pointers: [UnsafeRawPointer?]) -> CFBag? {
    var callbacks = kCFTypeBagCallBacks
    var pointers = pointers
    return pointers.withUnsafeMutableBufferPointer { buffer in
        CFBagCreate(nil, buffer.baseAddress, buffer.count, &callbacks)
    }
}

@_cdecl("cf_array_get_type_id")
public func cf_array_get_type_id() -> Int {
    Int(CFArrayGetTypeID())
}

@_cdecl("cf_array_create")
public func cf_array_create(_ values: UnsafePointer<UnsafeMutableRawPointer?>?, _ count: Int) -> UnsafeMutableRawPointer? {
    let pointers = UnsafeBufferPointer(start: values, count: count).map { raw -> UnsafeRawPointer? in
        raw.map(UnsafeRawPointer.init)
    }
    guard let array = acfArrayFromRawPointers(pointers) else { return nil }
    return Unmanaged.passRetained(array).toOpaque()
}

@_cdecl("cf_array_get_count")
public func cf_array_get_count(_ value: UnsafeMutableRawPointer) -> Int {
    let array = Unmanaged<CFArray>.fromOpaque(value).takeUnretainedValue()
    return CFArrayGetCount(array)
}

@_cdecl("cf_array_get_value_at_index")
public func cf_array_get_value_at_index(_ value: UnsafeMutableRawPointer, _ index: Int) -> UnsafeMutableRawPointer? {
    let array = Unmanaged<CFArray>.fromOpaque(value).takeUnretainedValue()
    guard index >= 0, index < CFArrayGetCount(array) else { return nil }
    let raw = CFArrayGetValueAtIndex(array, index)
    return acfRetainedCFType(UnsafeMutableRawPointer(mutating: raw))
}

@_cdecl("cf_dictionary_get_type_id")
public func cf_dictionary_get_type_id() -> Int {
    Int(CFDictionaryGetTypeID())
}

@_cdecl("cf_dictionary_create")
public func cf_dictionary_create(
    _ keys: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ values: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ count: Int
) -> UnsafeMutableRawPointer? {
    let keyPointers = UnsafeBufferPointer(start: keys, count: count).map { $0.map(UnsafeRawPointer.init) }
    let valuePointers = UnsafeBufferPointer(start: values, count: count).map { $0.map(UnsafeRawPointer.init) }
    guard let dictionary = acfDictionaryFromRawPointers(keys: keyPointers, values: valuePointers) else {
        return nil
    }
    return Unmanaged.passRetained(dictionary).toOpaque()
}

@_cdecl("cf_dictionary_get_count")
public func cf_dictionary_get_count(_ value: UnsafeMutableRawPointer) -> Int {
    let dictionary = Unmanaged<CFDictionary>.fromOpaque(value).takeUnretainedValue()
    return CFDictionaryGetCount(dictionary)
}

@_cdecl("cf_dictionary_contains_key")
public func cf_dictionary_contains_key(_ value: UnsafeMutableRawPointer, _ key: UnsafeMutableRawPointer) -> Bool {
    let dictionary = Unmanaged<CFDictionary>.fromOpaque(value).takeUnretainedValue()
    return CFDictionaryContainsKey(dictionary, UnsafeRawPointer(key))
}

@_cdecl("cf_dictionary_get_value")
public func cf_dictionary_get_value(_ value: UnsafeMutableRawPointer, _ key: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer? {
    let dictionary = Unmanaged<CFDictionary>.fromOpaque(value).takeUnretainedValue()
    let raw = CFDictionaryGetValue(dictionary, UnsafeRawPointer(key))
    return acfRetainedCFType(UnsafeMutableRawPointer(mutating: raw))
}

@_cdecl("cf_dictionary_copy_keys")
public func cf_dictionary_copy_keys(_ value: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer? {
    let dictionary = Unmanaged<CFDictionary>.fromOpaque(value).takeUnretainedValue()
    let count = CFDictionaryGetCount(dictionary)
    var keys = Array<UnsafeRawPointer?>(repeating: nil, count: count)
    CFDictionaryGetKeysAndValues(dictionary, &keys, nil)
    guard let array = acfArrayFromRawPointers(keys) else { return nil }
    return Unmanaged.passRetained(array).toOpaque()
}

@_cdecl("cf_dictionary_copy_values")
public func cf_dictionary_copy_values(_ value: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer? {
    let dictionary = Unmanaged<CFDictionary>.fromOpaque(value).takeUnretainedValue()
    let count = CFDictionaryGetCount(dictionary)
    var values = Array<UnsafeRawPointer?>(repeating: nil, count: count)
    CFDictionaryGetKeysAndValues(dictionary, nil, &values)
    guard let array = acfArrayFromRawPointers(values) else { return nil }
    return Unmanaged.passRetained(array).toOpaque()
}

@_cdecl("cf_bag_get_type_id")
public func cf_bag_get_type_id() -> Int {
    Int(CFBagGetTypeID())
}

@_cdecl("cf_bag_create")
public func cf_bag_create(_ values: UnsafePointer<UnsafeMutableRawPointer?>?, _ count: Int) -> UnsafeMutableRawPointer? {
    let pointers = UnsafeBufferPointer(start: values, count: count).map { $0.map(UnsafeRawPointer.init) }
    guard let bag = acfBagFromRawPointers(pointers) else { return nil }
    return Unmanaged.passRetained(bag).toOpaque()
}

@_cdecl("cf_bag_get_count")
public func cf_bag_get_count(_ value: UnsafeMutableRawPointer) -> Int {
    let bag = Unmanaged<CFBag>.fromOpaque(value).takeUnretainedValue()
    return CFBagGetCount(bag)
}

@_cdecl("cf_bag_contains_value")
public func cf_bag_contains_value(_ value: UnsafeMutableRawPointer, _ candidate: UnsafeMutableRawPointer) -> Bool {
    let bag = Unmanaged<CFBag>.fromOpaque(value).takeUnretainedValue()
    return CFBagContainsValue(bag, UnsafeRawPointer(candidate))
}

@_cdecl("cf_bag_get_count_of_value")
public func cf_bag_get_count_of_value(_ value: UnsafeMutableRawPointer, _ candidate: UnsafeMutableRawPointer) -> Int {
    let bag = Unmanaged<CFBag>.fromOpaque(value).takeUnretainedValue()
    return CFBagGetCountOfValue(bag, UnsafeRawPointer(candidate))
}

@_cdecl("cf_attributed_string_get_type_id")
public func cf_attributed_string_get_type_id() -> Int {
    Int(CFAttributedStringGetTypeID())
}

@_cdecl("cf_attributed_string_create")
public func cf_attributed_string_create(_ string: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer? {
    let string = Unmanaged<CFString>.fromOpaque(string).takeUnretainedValue()
    guard let attributed = CFAttributedStringCreate(nil, string, nil) else { return nil }
    return Unmanaged.passRetained(attributed).toOpaque()
}

@_cdecl("cf_attributed_string_get_string")
public func cf_attributed_string_get_string(_ value: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer? {
    let attributed = Unmanaged<CFAttributedString>.fromOpaque(value).takeUnretainedValue()
    return Unmanaged.passRetained(CFAttributedStringGetString(attributed)).toOpaque()
}

@_cdecl("cf_attributed_string_get_length")
public func cf_attributed_string_get_length(_ value: UnsafeMutableRawPointer) -> Int {
    let attributed = Unmanaged<CFAttributedString>.fromOpaque(value).takeUnretainedValue()
    return CFAttributedStringGetLength(attributed)
}

final class ACFTreeNode {
    let value: AnyObject?
    var children: [ACFTreeNode] = []

    init(value: AnyObject?) {
        self.value = value
    }
}

@_cdecl("cf_tree_create")
public func cf_tree_create(_ value: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    let object = value.map(acfBorrowedAnyObject)
    return Unmanaged.passRetained(ACFTreeNode(value: object)).toOpaque()
}

@_cdecl("cf_tree_append_child")
public func cf_tree_append_child(_ parent: UnsafeMutableRawPointer, _ child: UnsafeMutableRawPointer) {
    let parentNode = Unmanaged<ACFTreeNode>.fromOpaque(parent).takeUnretainedValue()
    let childNode = Unmanaged<ACFTreeNode>.fromOpaque(child).takeUnretainedValue()
    parentNode.children.append(childNode)
}

@_cdecl("cf_tree_get_child_count")
public func cf_tree_get_child_count(_ value: UnsafeMutableRawPointer) -> Int {
    let node = Unmanaged<ACFTreeNode>.fromOpaque(value).takeUnretainedValue()
    return node.children.count
}

@_cdecl("cf_tree_get_child_at_index")
public func cf_tree_get_child_at_index(_ value: UnsafeMutableRawPointer, _ index: Int) -> UnsafeMutableRawPointer? {
    let node = Unmanaged<ACFTreeNode>.fromOpaque(value).takeUnretainedValue()
    guard index >= 0, index < node.children.count else { return nil }
    return Unmanaged.passRetained(node.children[index]).toOpaque()
}

@_cdecl("cf_tree_copy_value")
public func cf_tree_copy_value(_ value: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer? {
    let node = Unmanaged<ACFTreeNode>.fromOpaque(value).takeUnretainedValue()
    guard let object = node.value else { return nil }
    return Unmanaged.passRetained(object).toOpaque()
}

File: src/cf/collections.rs
//! Core Foundation collection wrappers.
//!
#![allow(clippy::missing_panics_doc)]

//! ```rust
//! use apple_cf::cf::{CFArray, CFAttributedString, CFBag, CFDictionary, CFString, CFTree};
//!
//! let first = CFString::new("first");
//! let second = CFString::new("second");
//! let array = CFArray::from_values(&[&first, &second]);
//! assert_eq!(array.len(), 2);
//!
//! let dict = CFDictionary::from_pairs(&[(&first, &second)]);
//! assert!(dict.contains_key(&first));
//!
//! let bag = CFBag::from_values(&[&first, &first, &second]);
//! assert_eq!(bag.count_of_value(&first), 2);
//!
//! let attributed = CFAttributedString::new(&first);
//! assert_eq!(attributed.string().to_string(), "first");
//!
//! let root = CFTree::new(Some(&first));
//! let child = CFTree::new(Some(&second));
//! root.append_child(&child);
//! assert_eq!(root.child_count(), 1);
//! ```

use super::base::{impl_cf_type_wrapper, AsCFType, CFType, SwiftObject};
use super::CFString;
use crate::ffi;
use std::fmt;

impl_cf_type_wrapper!(CFArray, cf_array_get_type_id);
impl_cf_type_wrapper!(CFDictionary, cf_dictionary_get_type_id);
pub type CFDict = CFDictionary;
impl_cf_type_wrapper!(CFBag, cf_bag_get_type_id);
impl_cf_type_wrapper!(CFAttributedString, cf_attributed_string_get_type_id);

impl CFArray {
    /// Create an array from borrowed Core Foundation values.
    #[must_use]
    pub fn from_values(values: &[&dyn AsCFType]) -> Self {
        let raw_values: Vec<*mut std::ffi::c_void> = values.iter().map(|value| value.as_ptr()).collect();
        let ptr = unsafe { ffi::cf_array_create(raw_values.as_ptr(), raw_values.len()) };
        Self::from_raw(ptr).expect("CFArrayCreate returned NULL")
    }

    /// Number of elements in the array.
    #[must_use]
    pub fn len(&self) -> usize {
        unsafe { ffi::cf_array_get_count(self.as_ptr()) }
    }

    /// Whether the array is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Copy the value at `index`.
    #[must_use]
    pub fn get(&self, index: usize) -> Option<CFType> {
        let ptr = unsafe { ffi::cf_array_get_value_at_index(self.as_ptr(), index) };
        CFType::from_raw(ptr)
    }

    /// Copy all elements into a Rust vector.
    #[must_use]
    pub fn values(&self) -> Vec<CFType> {
        (0..self.len()).filter_map(|index| self.get(index)).collect()
    }
}

impl CFDictionary {
    /// Create a dictionary from borrowed key/value pairs.
    #[must_use]
    pub fn from_pairs(pairs: &[(&dyn AsCFType, &dyn AsCFType)]) -> Self {
        let keys: Vec<*mut std::ffi::c_void> = pairs.iter().map(|(key, _)| key.as_ptr()).collect();
        let values: Vec<*mut std::ffi::c_void> = pairs.iter().map(|(_, value)| value.as_ptr()).collect();
        let ptr = unsafe { ffi::cf_dictionary_create(keys.as_ptr(), values.as_ptr(), pairs.len()) };
        Self::from_raw(ptr).expect("CFDictionaryCreate returned NULL")
    }

    /// Number of key/value pairs.
    #[must_use]
    pub fn len(&self) -> usize {
        unsafe { ffi::cf_dictionary_get_count(self.as_ptr()) }
    }

    /// Whether the dictionary is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Whether `key` exists in the dictionary.
    #[must_use]
    pub fn contains_key(&self, key: &dyn AsCFType) -> bool {
        unsafe { ffi::cf_dictionary_contains_key(self.as_ptr(), key.as_ptr()) }
    }

    /// Copy the value associated with `key`.
    #[must_use]
    pub fn get(&self, key: &dyn AsCFType) -> Option<CFType> {
        let ptr = unsafe { ffi::cf_dictionary_get_value(self.as_ptr(), key.as_ptr()) };
        CFType::from_raw(ptr)
    }

    /// Copy all keys into a `CFArray`.
    #[must_use]
    pub fn keys(&self) -> CFArray {
        let ptr = unsafe { ffi::cf_dictionary_copy_keys(self.as_ptr()) };
        CFArray::from_raw(ptr).expect("CFDictionary keys array should be non-null")
    }

    /// Copy all values into a `CFArray`.
    #[must_use]
    pub fn values(&self) -> CFArray {
        let ptr = unsafe { ffi::cf_dictionary_copy_values(self.as_ptr()) };
        CFArray::from_raw(ptr).expect("CFDictionary values array should be non-null")
    }
}

impl CFBag {
    /// Create a bag from borrowed Core Foundation values.
    #[must_use]
    pub fn from_values(values: &[&dyn AsCFType]) -> Self {
        let raw_values: Vec<*mut std::ffi::c_void> = values.iter().map(|value| value.as_ptr()).collect();
        let ptr = unsafe { ffi::cf_bag_create(raw_values.as_ptr(), raw_values.len()) };
        Self::from_raw(ptr).expect("CFBagCreate returned NULL")
    }

    /// Number of values in the bag.
    #[must_use]
    pub fn len(&self) -> usize {
        unsafe { ffi::cf_bag_get_count(self.as_ptr()) }
    }

    /// Whether the bag is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Whether the bag contains `candidate`.
    #[must_use]
    pub fn contains(&self, candidate: &dyn AsCFType) -> bool {
        unsafe { ffi::cf_bag_contains_value(self.as_ptr(), candidate.as_ptr()) }
    }

    /// Number of times `candidate` appears in the bag.
    #[must_use]
    pub fn count_of_value(&self, candidate: &dyn AsCFType) -> usize {
        unsafe { ffi::cf_bag_get_count_of_value(self.as_ptr(), candidate.as_ptr()) }
    }
}

impl CFAttributedString {
    /// Create an attributed string with no attributes.
    #[must_use]
    pub fn new(string: &CFString) -> Self {
        let ptr = unsafe { ffi::cf_attributed_string_create(string.as_ptr()) };
        Self::from_raw(ptr).expect("CFAttributedStringCreate returned NULL")
    }

    /// Underlying plain string.
    #[must_use]
    pub fn string(&self) -> CFString {
        let ptr = unsafe { ffi::cf_attributed_string_get_string(self.as_ptr()) };
        CFString::from_raw(ptr).expect("CFAttributedStringGetString returned NULL")
    }

    /// Character length.
    #[must_use]
    pub fn len(&self) -> usize {
        unsafe { ffi::cf_attributed_string_get_length(self.as_ptr()) }
    }

    /// Whether the string is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

/// Safe wrapper around a Swift-backed `CFTree` helper.
#[derive(Clone, PartialEq, Eq, Hash)]
pub struct CFTree(SwiftObject);

impl CFTree {
    /// Create a tree node with an optional Core Foundation payload.
    #[must_use]
    pub fn new(value: Option<&dyn AsCFType>) -> Self {
        let ptr = unsafe { ffi::cf_tree_create(value.map_or(std::ptr::null_mut(), AsCFType::as_ptr)) };
        Self(SwiftObject::from_raw(ptr).expect("tree bridge returned NULL"))
    }

    #[must_use]
    pub(crate) fn from_raw(ptr: *mut std::ffi::c_void) -> Option<Self> {
        SwiftObject::from_raw(ptr).map(Self)
    }

    /// Borrow the raw tree handle.
    #[must_use]
    pub(crate) const fn as_ptr(&self) -> *mut std::ffi::c_void {
        self.0.as_ptr()
    }

    /// Append `child` to this node.
    pub fn append_child(&self, child: &Self) {
        unsafe { ffi::cf_tree_append_child(self.as_ptr(), child.as_ptr()) };
    }

    /// Number of direct children.
    #[must_use]
    pub fn child_count(&self) -> usize {
        unsafe { ffi::cf_tree_get_child_count(self.as_ptr()) }
    }

    /// Copy the child at `index`.
    #[must_use]
    pub fn child_at(&self, index: usize) -> Option<Self> {
        let ptr = unsafe { ffi::cf_tree_get_child_at_index(self.as_ptr(), index) };
        Self::from_raw(ptr)
    }

    /// Copy the payload value if present.
    #[must_use]
    pub fn value(&self) -> Option<CFType> {
        let ptr = unsafe { ffi::cf_tree_copy_value(self.as_ptr()) };
        CFType::from_raw(ptr)
    }
}

impl fmt::Debug for CFTree {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("CFTree")
            .field("ptr", &self.as_ptr())
            .field("child_count", &self.child_count())
            .finish()
    }
}
